### ISSUE TITLE
Update default `http_ssl_version`

### ISSUE THREAD

**Issue by eafif** [wildbit/postmark-gem#86]
Getting the following error with the latest version of openssl on Ubuntu 20.04:
`OpenSSL::SSL::SSLError: SSL_connect returned=1 errno=0 state=error: no protocols available`

```
apt-cache policy openssl
openssl:
  Installed: 1.1.1f-1ubuntu2
  Candidate: 1.1.1f-1ubuntu2
```

I can get it to work by using `Postmark::ApiClient.new(api_token, {http_ssl_version: :TLSv1_2})`.

Calling `openssl s_client -connect api.postmarkapp.com:443` shows that the certificate uses `TLSv1.2`:

```
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
```

**Comment by kot-begemot** [comment]
Same thing there. I got exactly same result and can confirm that TSLv1 works on ubuntu 18.04 and does not work on ubuntu 20.04. Here is some more info for research

My local machine (Ubuntu 18.04)
```
$ curl -Lks 'https://git.io/rg-ssl' | ruby
Here's your Ruby and OpenSSL environment:

Ruby:           2.6.6p146 (2020-03-31 revision 67876) [x86_64-linux]
RubyGems:       3.0.8
Bundler:        1.17.3
Compiled with:  OpenSSL 1.1.1d  10 Sep 2019
Loaded version: OpenSSL 1.1.1  11 Sep 2018
SSL_CERT_FILE:  /usr/lib/ssl/cert.pem
SSL_CERT_DIR:   /usr/lib/ssl/certs

With that out of the way, let's see if you can connect to rubygems.org...

Bundler connection to rubygems.org:       success ✅
RubyGems connection to rubygems.org:      success ✅
Ruby net/http connection to rubygems.org: success ✅

Hooray! This Ruby can connect to rubygems.org. You are all set to use Bundler and RubyGems. 👌
```

My remote machine (Ubuntu 20.04)
```
$ curl -Lks 'https://git.io/rg-ssl' | ruby
Here's your Ruby and OpenSSL environment:

Ruby:           2.6.6p146 (2020-03-31 revision 67876) [x86_64-linux]
RubyGems:       3.0.3
Bundler:        2.1.4
Compiled with:  OpenSSL 1.1.1f  31 Mar 2020
Loaded version: OpenSSL 1.1.1f  31 Mar 2020
SSL_CERT_FILE:  /usr/lib/ssl/cert.pem
SSL_CERT_DIR:   /usr/lib/ssl/certs

With that out of the way, let's see if you can connect to rubygems.org...

Bundler connection to rubygems.org:       success ✅
RubyGems connection to rubygems.org:      success ✅
Ruby net/http connection to rubygems.org: success ✅

Hooray! This Ruby can connect to rubygems.org. You are all set to use Bundler and RubyGems. 👌
```

Here is a script you can use to confirm this issue in console
```
require "net/http"

http = Net::HTTP::Proxy(nil, nil, nil, nil).new("api.postmarkapp.com", 443)

      http.read_timeout = 15
      http.open_timeout = 5
      http.use_ssl = true
      http.ssl_version = :TLSv1_2
      http

 headers = { 'X-Postmark-Server-Token' => '6537357a-1d37-4aae-b903-b3b5b6c5f744' }

http.post('/', "", headers)
```

**Comment by mbbernstein** [comment]
Same here
```
connect_nonblock': SSL_connect returned=1 errno=0 state=error: no protocols available (OpenSSL::SSL::SSLError)
```

**Comment by mbbernstein** [comment]
The solution for this is here:
https://askubuntu.com/questions/1233186/ubuntu-20-04-how-to-set-lower-ssl-security-level

**Comment by ibalosh** [comment]
Hi guys

@mbbernstein thank you for the link to solution of the issue. We will make sure to add this to our [wiki pages](https://github.com/wildbit/postmark-gem/wiki/Known-issues-and-how-to-resolve-them)

**Comment by mattbrictson** [comment]
Ideally we shouldn't need to employ a workaround to connect using a less secure configuration. `api.postmarkapp.com` currently gets a "B" grade for SSL support, which could be improved: https://www.ssllabs.com/ssltest/analyze.html?d=api.postmarkapp.com

**Comment by sergiopantoja** [comment]
Just ran into the same issue. Agreed that this workaround shouldn't have to be made. I'm using [postmark-rails](https://github.com/wildbit/postmark-rails), so I'm passing the `http_ssl_version` option into my `config/application.rb` and it's now working as expected. Thanks @eafif 

```ruby
config.action_mailer.postmark_settings = {
    api_token: 'xxx',
    http_ssl_version: :TLSv1_2
  }
```

**Comment by mrthan** [comment]
@ibalosh this needs attention. Just ran across this issue in WSL2 Ubuntu environment, nothing in the wiki yet. Thanks @sergiopantoja your solution seems the best fit until the PR is merged.

**Comment by ibalosh** [comment]
Hi @mrthan 

there is a pull request by @nickhammond  on our team which should resolve this one soon: https://github.com/wildbit/postmark-gem/pull/92

Although it's a small change in the codebase it's affecting all users of the gem, so we need bit more time until will release it.
Until then, ssl version can be updated in current gem on client initialization like @sergiopantoja described.

**Comment by calebfaruki** [comment]
This error also affects Heroku cedar stack 20, which runs on Ubuntu 20.04.

```
OpenSSL::SSL::SSLError (SSL_connect returned=1 errno=0 state=error: no protocols available)
```

**Comment by edmorley** [comment]
@ibalosh @nickhammond @tomazy Hi from Heroku!

We're starting to get reports of this error from Heroku customers who have upgraded to the [recently GAed](https://devcenter.heroku.com/changelog-items/1951) Heroku-20 stack, which is based upon Ubuntu 20.04.

In one of those reports the customer mentioned Postmark support were telling users to adjust their global Ubuntu openssl configuration to use insecure cyphers (with them having been given only a link to https://askubuntu.com/questions/1233186/ubuntu-20-04-how-to-set-lower-ssl-security-level) and didn't mention that it was actually a Postmark bug.

Adjusting the global openssl configuration is really bad from a security point of view (so shouldn't be recommended to customers), and is also not possible on environments where the user does not have root access (such as Heroku).

Please could the fix in #92 be released as soon as possible? 

And in the meantime, could Postmark support please give app-config level workarounds (and not global openssl config workarounds) such as those listed in https://github.com/wildbit/postmark-gem/issues/86#issuecomment-669736462 ?

Many thanks :-)

**Comment by ibalosh** [comment]
Hey @edmorley @calebfaruki 

@nickhammond will check it out soon, until then, doesn't the solution one of the users above mentioned work:

Postmark::ApiClient.new(api_token, {http_ssl_version: :TLSv1_2})

to set different version of tls. It's not a bug in the gem, it is only not defaulting yet to newer TLS version since this would affect many other users too and might break things for them.

**Comment by edmorley** [comment]
Hi! Thank you for the quick reply and additional workaround using this gem rather than the rails one.

> It's not a bug in the gem, but rather it's not defaulting yet to newer TLS version since this would affect many other users too and might break things for them.

The TLSv1 (1.0) protocol is insecure - to the extent that browsers have recently dropped support for it, and much of the ecosystem did several years ago, eg:
https://blog.rubygems.org/2018/02/24/tls-10-and-11-deprecation-notice.html

The biggest concern I have is that this gem hardcodes the TLSv1 version, rather than permitting a range of versions. If the gem set correct `max_version` and `min_version` on the `Net::HTTP` usages (or say left them at defaults) then the clients could negotiate the appropriate protocol version.

**Comment by jarthod** [comment]
I confirm this fix:
```ruby
Postmark::ApiClient.new(api_token, {http_ssl_version: :TLSv1_2})
```
Works as I am currently using it.

> The biggest concern I have is that this gem hardcodes the TLSv1 version, rather than permitting a range of versions. If the gem set correct max_version and min_version on the Net::HTTP usages (or say left them at defaults) then the clients could negotiate the appropriate protocol version.

I cannot agree more, Indeed if you force TLSv1.2 it has more chance to break than just removing this parameter and letting the HTTP client choosing the best protocol. Why not do this instead? You can always keep the option to specify a TLS version in the code, and let people know in the changelog that if they have problem with the upgrade (which would be caused by their own setup), they can always specify a hardcoded version.

**Comment by henrahmagix** [comment]
As above, I would like to set `http_ssl_version: nil`, but unfortunately that is ignored by this line:
https://github.com/wildbit/postmark-gem/blob/a596f6e8ed47ec657233cc42454dc728f43cbb85/lib/postmark/http_client.rb#L59

For the meantime, setting explicitly `http_ssl_version: :TLSv1_2` works, but I'd prefer not to have to look after that value in the future.

I've added https://github.com/wildbit/postmark-gem/pull/96 to allow `http_ssl_version: nil` =)

**Comment by tonydehnke** [comment]
Just spent hours chasing this one down, I use the `postmark-rails` gem in my app hosted on Heroku.

I was able to get it working on Heroku-20 Buildpack using the info above from @sergiopantoja (THANK YOU!), incase it helps anyone here is what my line looks like when using and ENV for the Postmark token:

`config.action_mailer.postmark_settings = { api_token: ENV['POSTMARK_TOKEN'], http_ssl_version: :TLSv1_2 }`

**Comment by edmorley** [comment]
@ibalosh @nickhammond Hi! We've just had further support tickets. Please could the fix be released this week?

**Comment by nickhammond** [comment]
@edmorley Yep! We'll be merging in #92 and releasing an update for that, apologies for the delay.

**Comment by edmorley** [comment]
@nickhammond Thank you :-)

**Comment by tonydehnke** [comment]
@nickhammond  will we see the fix In the Postmark-Rails gem too?  https://github.com/wildbit/postmark-rails/issues/82

**Comment by nickhammond** [comment]
Hi all,

Apologies for the issues with SSL, I've just released a new version of the postmark gem(1.21.3) that no longer supplies a default SSL version and instead uses the default provided via OpenSSL. You can still specify a specific version if you'd like but it's no longer required.  Let us know if you have any issues, going to close this for now but feel free to re-open it if needed.

**Comment by edmorley** [comment]
@nickhammond Thank you! :-D
